sccache-dist/src/cache.rs

```diff
@@ -0,0 +1,158 @@
+use lru_disk_cache::{LruDiskCache, ReadSeek};
+use lru_disk_cache::Error as LruError;
+use lru_disk_cache::Result as LruResult;
+use ring::digest::{SHA512, Context};
+use serde_json;
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::fs::{self, File};
+use std::io::{self, BufReader, Read, Write};
+use std::path::{Path, PathBuf};
+use std::sync::Mutex;
+
+use errors::*;
+
+// TODO: possibly shouldn't be public
+pub struct ClientToolchainCache {
+    cache_dir: PathBuf,
+    cache: Mutex<TcCache>,
+    // Local machine mapping from 'weak' hashes to strong toolchain hashes
+    weak_map: Mutex<HashMap<String, String>>,
+}
+
+impl ClientToolchainCache {
+    pub fn new(cache_dir: &Path, cache_size: u64) -> Self {
+        let cache_dir = cache_dir.to_owned();
+        fs::create_dir_all(&cache_dir).unwrap();
+
+        let weak_map_path = cache_dir.join("weak_map.json");
+        if !weak_map_path.exists() {
+            fs::File::create(&weak_map_path).unwrap().write_all(b"{}").unwrap()
+        }
+        let weak_map = serde_json::from_reader(fs::File::open(weak_map_path).unwrap()).unwrap();
+
+        let tc_cache_dir = cache_dir.join("tc");
+        let cache = Mutex::new(TcCache::new(&tc_cache_dir, cache_size).unwrap());
+
+        Self {
+            cache_dir,
+            cache,
+            // TODO: shouldn't clear on restart, but also should have some
+            // form of pruning
+            weak_map: Mutex::new(weak_map),
+        }
+    }
+
+    // Get the bytes of a toolchain tar
+    pub fn get_toolchain_cache(&self, key: &str) -> Option<Vec<u8>> {
+        let mut toolchain_reader = match self.cache.lock().unwrap().get(key) {
+            Ok(rdr) => rdr,
+            Err(LruError::FileNotInCache) => return None,
+            Err(e) => panic!("{}", e),
+        };
+        let mut ret = vec![];
+        toolchain_reader.read_to_end(&mut ret).unwrap();
+        Some(ret)
+    }
+    // TODO: It's more correct to have a FnBox or Box<FnOnce> here
+    // If the toolchain doesn't already exist, create it and insert into the cache
+    pub fn put_toolchain_cache(&self, weak_key: &str, create: &mut FnMut(fs::File) -> io::Result<()>) -> Result<String> {
+        if let Some(strong_key) = self.weak_to_strong(weak_key) {
+            debug!("Using cached toolchain {} -> {}", weak_key, strong_key);
+            return Ok(strong_key)
+        }
+        debug!("Weak key {} appears to be new", weak_key);
+        // TODO: don't use this as a poor exclusive lock on this global file
+        let mut cache = self.cache.lock().unwrap();
+        let file = fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .open("/tmp/toolchain_cache.tar")?;
+        create(file)?;
+        // TODO: after, if still exists, remove it
```

Comment:
This sounds like something that should definitely be fixed.

Reply:
Refined caching and scheduling are a slightly later PR.